src/parquet/ReadParquet.cpp

```diff
@@ -1,5 +1,407 @@
+    }
+
+    std::vector<int64_t> startIdxPerCol;
+    startIdxPerCol.resize(num_cols, startIdx);
+
+    int64_t row_idx = 0;
+
+    for (int rg_idx = 0; (rg_idx<num_row_groups) && (row_idx<numElems);
+         rg_idx++) {
+
+      std::shared_ptr<parquet::RowGroupReader> row_group_reader =
+          parquet_reader->RowGroup(rg_idx);
+
+      std::shared_ptr<parquet::ColumnReader> column_reader;
+
+      int64_t nrows_in_iter = -1;
+
+      for (int col_idx = 0; col_idx<num_cols ; col_idx++) {
+        column_reader = row_group_reader->Column(col_idx);
+        const parquet::ColumnDescriptor* col_info =
+          file_metadata->schema()->Column(col_idx);
+
+        void* chpl_arr = chpl_arrs[col_idx];
+
+        // TODO, I want values of this type to be created per read operation,
+        // not per column, per rowgroup
+        auto op = AkColReadOp { chpl_arr,
+                                &startIdxPerCol[col_idx],
+                                column_reader,
+                                false, // has_non_float_nulls is for backward
+                                       // compat and ignored while reading all
+                                       // columns. nullMode is used instead.
+                                nullMode,
+                                row_idx,
+                                numElems,
+                                batchSize,
+                                where_null_chpl,
+                                col_info };
+
+        int64_t nread;
+        switch (types[col_idx]) {
+          case ARROWFLOAT:   nread = op.read<arrow::FloatType>();      break;
+          case ARROWDOUBLE:  nread = op.read<arrow::DoubleType>();     break;
+          case ARROWINT64:   nread = op.read<arrow::Int64Type>();      break;
+          case ARROWUINT64:  nread = op.read<arrow::UInt64Type>();     break;
+          case ARROWBOOLEAN: nread = op.read<arrow::BooleanType>();    break;
+          case ARROWINT32:   nread = op.read<arrow::Int32Type>();      break;
+          case ARROWUINT32:  nread = op.read<arrow::UInt32Type>();     break;
+          case ARROWDECIMAL: nread = op.read<arrow::Decimal128Type>(); break;
+          default:
+            std::cout << "\tUnknown Arrow type\n";
+            return ARROWERROR;
+        }
+
+        if (nrows_in_iter == -1) {
+          // this is the first time we are reading in this row group. We'll need
+          // to bump up the row_idx in the end of the iteration by nrows_read
+          nrows_in_iter = nread;
+        }
+        else {
+          // we already now how many rows we are reading per column, which is
+          // stored in nrows_in_iter. But did we read the correct number of rows
+          // in this particular iteration?
+          if (nread != nrows_in_iter) {
+            std::cout << "\tUneven number of rows are read.";
+            std::cout << " Expected " << nrows_in_iter;
+            std::cout << " But read " << nread << " instead\n";
```

Comment:
nit, these errors should go to `std::cerr`

Reply:
Good catch. We populate errMsg based on the exceptions's message at the end of the function. I just decided to throw `std::exception`s here and above to be captured at the end. I think we can consider tighter error handling schemes on the C++ side in the future. I added a TODO comment.